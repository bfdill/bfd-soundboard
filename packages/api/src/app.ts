import * as Koa from 'koa'
import * as KoaHelmet from 'koa-helmet'
import * as KoaBody from 'koa-body'
import * as KoaRouter from 'koa-router'
import * as cors from '@koa/cors'
import * as winston from 'winston'
import { createLogger } from '@talkboy-9000/logging'
import { getHealthcheckRouter } from './controllers/healthcheck'
import { getPlayerRouter } from './controllers/player'
import { getSoundsRouter } from './controllers/sounds'
import {
  ApplicationRouter,
  ApplicationKoa,
  ApplicationState,
  IApplication,
  IApplicationContext
} from './types'
import { ISystemMiddleware, SystemMiddleware } from './middleware'

export const configureRouter = (router: ApplicationRouter) =>
  router
    .use('/healthcheck', getHealthcheckRouter().routes())
    .use('/player', getPlayerRouter().routes())
    .use('/sounds', getSoundsRouter().routes())

export const getAppLogger = () =>
  createLogger({
    app: {
      // uhh, got config
      name: 'talkboy-9000',
      version: '2.3.6'
    }
  })
export const getAppRouter = () =>
  configureRouter(new KoaRouter<ApplicationState, IApplicationContext>())

export class App implements IApplication {
  private static instance: IApplication | undefined

  constructor(
    protected readonly koa: ApplicationKoa,
    protected readonly router: ApplicationRouter,
    protected readonly logger: winston.Logger,
    protected readonly systemMiddleware: ISystemMiddleware
  ) {}

  protected setup = (
    koa: ApplicationKoa,
    router: ApplicationRouter,
    logger: winston.Logger,
    systemMiddleware: ISystemMiddleware
  ) => {
    koa.on('error', err => {
      logger.error(err)
    })

    koa
      .use(cors())
      .use(KoaHelmet())
      .use(KoaBody())
      .use(systemMiddleware.middleware)
      .use(router.routes())
      .use(router.allowedMethods())
  }

  listen = (port: number) => {
    this.logger.info({ listen: port })
    this.setup(this.koa, this.router, this.logger, this.systemMiddleware)

    return this.koa.listen(port)
  }

  static getInstance(): IApplication {
    if (this.instance !== undefined) return this.instance

    const logger = getAppLogger()

    this.instance = new App(
      new Koa<ApplicationState, IApplicationContext>(),
      getAppRouter(),
      logger,
      new SystemMiddleware(logger)
    )

    return this.instance
  }
}
