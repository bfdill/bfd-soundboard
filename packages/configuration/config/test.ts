import { DeepPartial } from '@talkboy-9000/utils'

import { FullConfig } from '../src/configuration.types'

const config: DeepPartial<FullConfig> = {}

export default config
