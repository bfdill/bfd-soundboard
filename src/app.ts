import * as Koa from 'koa'
import * as KoaHelmet from 'koa-helmet'
import * as KoaBody from 'koa-body'
import * as KoaRouter from 'koa-router'
import * as winston from 'winston'
import {
  createLogger,
  ILoggingMiddleware,
  LoggingMiddleware
} from './modules/logging'
import { getHealthcheckRouter } from './controllers/healthcheck'
import { getPlayerRouter } from './controllers/player'
import { soundsRouter } from './controllers/sounds'
import {
  ApplicationRouter,
  ApplicationKoa,
  ApplicationState,
  IApplicationContext,
  IApplication
} from './types'

export const configureRouter = (router: ApplicationRouter) =>
  router
    .use('/healthcheck', getHealthcheckRouter().routes())
    .use('/player', getPlayerRouter().routes())
    .use('/sounds', soundsRouter.routes())
export const getAppLogger = () =>
  createLogger({
    app: {
      // uhh, got config
      name: 'talkboy-9000',
      version: '2.3.6'
    }
  })
export const getAppRouter = () =>
  configureRouter(new KoaRouter<ApplicationState, IApplicationContext>())

export class App implements IApplication {
  private static instance: IApplication | undefined

  constructor(
    protected readonly koa: ApplicationKoa,
    protected readonly router: ApplicationRouter,
    protected readonly logger: winston.Logger,
    protected readonly loggingMiddleware: ILoggingMiddleware
  ) {}

  protected setup = (
    koa: ApplicationKoa,
    router: ApplicationRouter,
    logger: winston.Logger,
    loggingMiddleware: ILoggingMiddleware
  ) => {
    koa.on('error', err => {
      logger.error(err)
    })

    koa
      .use(KoaHelmet())
      .use(KoaBody())
      .use(loggingMiddleware.middleware)
      .use(router.routes())
      .use(router.allowedMethods())
  }

  listen = (port: number) => {
    this.logger.info(`listen(${port})`)
    this.setup(this.koa, this.router, this.logger, this.loggingMiddleware)

    return this.koa.listen(port)
  }

  static getInstance(): IApplication {
    if (this.instance !== undefined) return this.instance

    this.instance = new App(
      new Koa<ApplicationState, IApplicationContext>(),
      getAppRouter(),
      getAppLogger(),
      new LoggingMiddleware()
    )

    return this.instance
  }
}
