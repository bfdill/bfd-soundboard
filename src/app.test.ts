import * as Koa from 'koa'
import * as koaRouter from 'koa-router'
import * as winston from 'winston'
import { KoaApp, configureRouter } from './app'
import { inspect } from 'util'
import { getJestLogger } from './modules/winston-jest/index.test'

describe('app', () => {
  test('has expected exports', () => {
    expect(Object.keys(require('./app'))).toMatchSnapshot()
  })

  test('configureRouter', () => {
    const fakeRouter: koaRouter = {} as any
    const mockUse = jest.fn().mockReturnValue(fakeRouter)
    fakeRouter.use = mockUse

    configureRouter(fakeRouter)

    mockUse.mock.calls.forEach(call => {
      expect(call).toMatchSnapshot()
    })
  })

  describe('KoaApp', () => {
    const mockKoaUse = jest.fn()
    const mockKoa: Koa = {
      listen: jest.fn(),
      on: jest.fn(),
      use: mockKoaUse
    } as any
    const mockRouter: koaRouter = {
      allowedMethods: jest.fn(),
      routes: jest.fn()
    } as any

    const jestLogger = getJestLogger()

    const getSut = (
      koa: Koa = mockKoa,
      router: koaRouter = mockRouter,
      logger: winston.Logger = jestLogger.logger
    ) => new KoaApp(koa, router, logger)
    const TEST_PORT = 12345

    beforeEach(() => {
      mockKoaUse.mockReturnValue(mockKoa)

      const sut = getSut()
      sut.listen(TEST_PORT)
    })

    test('listen logs', () => {
      jestLogger.callsMatchSnapshot()
    })

    test('koa.use matches snapshot', () => {
      expect.assertions(4)
      mockKoaUse.mock.calls.forEach(call => {
        // this is like super meh 🦐
        expect(inspect(call)).toMatchSnapshot()
      })
    })

    test('calls routes', () => {
      expect(mockRouter.routes).toHaveBeenCalled()
    })

    test('calls allowedMethods', () => {
      expect(mockRouter.allowedMethods).toHaveBeenCalled()
    })

    test('on error is registered and logs an error', () => {
      const koaOn = (<jest.Mock>mockKoa.on).mock.calls[0]
      const error = new Error('can has error')

      expect(koaOn[0]).toEqual('error')
      koaOn[1](error)
      jestLogger.callsMatchSnapshot()
    })
  })
})
