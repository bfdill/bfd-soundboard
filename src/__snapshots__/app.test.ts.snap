// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app KoaApp koa.use matches snapshot 1`] = `"[ { [Function: middleware] _name: 'helmet' } ]"`;

exports[`app KoaApp koa.use matches snapshot 2`] = `"[ [Function] ]"`;

exports[`app KoaApp koa.use matches snapshot 3`] = `"[ undefined ]"`;

exports[`app KoaApp koa.use matches snapshot 4`] = `"[ undefined ]"`;

exports[`app KoaApp listen logs 1`] = `
Array [
  Object {
    "level": "info",
    "message": "listen(12345)",
    Symbol(level): "info",
    Symbol(message): "{\\"message\\":\\"listen(12345)\\",\\"level\\":\\"info\\"}",
  },
]
`;

exports[`app KoaApp on error is registered and logs an error 1`] = `
Array [
  Object {
    "level": "info",
    "message": "listen(12345)",
    Symbol(level): "info",
    Symbol(message): "{\\"message\\":\\"listen(12345)\\",\\"level\\":\\"info\\"}",
  },
]
`;

exports[`app KoaApp on error is registered and logs an error 2`] = `
Array [
  [Error: can has error],
]
`;

exports[`app configureRouter 1`] = `
Array [
  "/healthcheck",
  [Function],
]
`;

exports[`app configureRouter 2`] = `
Array [
  "/player",
  [Function],
]
`;

exports[`app configureRouter 3`] = `
Array [
  "/sounds",
  [Function],
]
`;

exports[`app has expected exports 1`] = `
Array [
  "configureRouter",
  "KoaApp",
  "router",
  "logger",
]
`;
